README.md
# IP creator (makeip)

**IP creator** (`makeip`) is an utility used for generating homebrew **Sega
Dreamcast** bootstrap files, also known as `IP` which stands for **Initial
Program**. This file is often named `IP.BIN`. It's a critical component for
any Dreamcast discs.

`IP.BIN` file is used for making selfboot (**MIL-CD**) discs for **Sega
Dreamcast** softwares. When the file is inserted in the bootsector of a disc,
the bootstrap program (`IP.BIN`) will execute the main program stored on the
file system which is often named `1ST_READ.BIN`. Please note, the `IP.BIN`
file is not needed on the file system as it's inserted in the bootsector.

**IP creator** will generates a custom `IP.BIN` file designed for homebrew
programs. It uses minimal copyrighted code, i.e. the minimum code that
can't be removed/altered in any way.

All `IP.BIN` files contains custom data fields used to describe some properties
of the software present on the disc. **IP creator** will allow you to edit
these fields. It also supports the insertion of custom images in the boot
screen. These kind of images are called **MR images** or **MR logos** and are
often stored in files with the `.mr` extension. You'll be able to convert
a **PNG** image into the **MR** format on-the-fly with **IP creator**.

Below you'll find an example of the *ADK/Napalm* MR image shown in the boot
screen (this **MR** logo was extracted from the `dcload` bootstrap):

![Example](img/mrlogo.png "Sample MR image")

## Building

This program is a standard C program which may be compiled with **GNU Make**.
It requires `libpng-dev` installed.
[Learn more about libpng here](http://www.libpng.org/pub/png/libpng.html).

1. Edit the `Makefile` and check if everything is OK for you (e.g. `libpng`
   directories);
2. Enter `make` (`gmake` on BSD systems).

## Usage

To use this tool, several modes are available:

1. Using a IP template file (`ip.txt`);
2. Using command-line arguments.

Available options are (displayed with the `-h` switch):
	
	-f                 Force overwrite output file if already exist
	-h                 Print usage information (you're looking at it)
	-l <infilename>    Load/insert an image into bootstrap (MR; PNG)
	-t <tmplfilename>  Use an external IP.TMPL file (override default)
	-u                 Print field usage information
	-s <outfilename>   Save image from <infilename> to MR format (see '-l')
	-v                 Enable verbose mode

To learn more about **MR images**, please read below. You may use either a
raw `MR` image or a `PNG` image that will be converted on-the-fly.

### Using a IP template file

This mode is the historical way to populate the fields of the `IP.BIN` file.

Create a text file called `ip.txt` and write this content on it:

	Device Info   : CD-ROM1/1
	Area Symbols  : JUE
	Peripherals   : E000F10
	Product No    : T0000
	Version       : V1.000
	Release Date  : 20000627
	Boot Filename : 1ST_READ.BIN
	SW Maker Name : YOUR NAME HERE
	Game Title    : TITLE OF THE SOFTWARE

Then run the following command:

	makeip ip.txt IP.BIN

This will generate the `IP.BIN` with the above values.

You don't have to fill all the fields, removing them from the `ip.txt` will use
the default. So a minimal `ip.txt` file may be like:

	Product No    : T0000
	Version       : V1.000
	SW Maker Name : YOUR NAME HERE
	Game Title    : TITLE OF THE SOFTWARE

**Note:** In the original `ip.txt` files, you may have the fields **Hardware
ID** and **Maker ID** declared. They can't be altered so it isn't necessary
anymore to pass them. Plus, the **Device Info** field may contains a fake
**CRC** like `0000` (see below in the **Device Info** field sub-section for
more info), this isn't necessary now but it will still work if present. 

###  Using command-line arguments

If you want you may pass directly the field information by using the proper
switch. To print this list, use the `-u` switch instead of `-h`.

	-a <areasymbols>    Area sym (J)apan, (U)SA, (E)urope (default: JUE)
	-b <bootfilename>   Boot filename (default: 1ST_READ.BIN)
	-c <companyname>    Company name / SW maker name (default: KallistiOS)
	-d <releasedate>    Release date (format: YYYYMMDD, default: <today>)
	-e <version>        Product version (default: V1.000)
	-g <gametitle>      Title of the software (default: GAMETITLE)
	-i <deviceinfo>     Device info (format: CD-ROMx/y, default: CD-ROM1/1)
	-n <productno>      Product number (default: T-00000)
	-p <peripherals>    Peripherals (default: E000F10)

## MR Images

**MR Image** is a special image format that can be inserted in the boostrap.
This kind of image are often saved as `iplogo.mr` or equivalent.

With **IP creator** you will be able to:

* Insert a custom image in the generated bootstrap. It can be either raw
  **MR** format or **PNG** format;
* Convert an input file (i.e. a **PNG** file) into the **MR** format and
  save the result on the disk for later use.

### Inserting an image into bootstrap

Just pass the `-l` switch to the command-line:

	makeip -l iplogo.mr -v IP.BIN

You may pass a **PNG** file too:

	makeip -l iplogo.png -v IP.BIN

In that case, the **PNG** file will be converted on-the-fly in the **MR**
format before inserting in the bootstrap.

### About MR Image constraints

To be used in the generated `IP.BIN` file, the **MR image** must be:

1. **320 * 90** or less;
2. Less than **128** colors;
3. Less than **8192 Bytes**.

The transparent color is `#c0c0c0`, or `192`, `192`, `192` in RGB.

### Converting a PNG Image into a MR Image

If you just want to convert a **PNG** Image to a **MR** Image but not
generating the bootstrap, you may use **IP creator** like:

	makeip -l iplogo.png -s iplogo.mr -v

This will try to convert the `iplogo.png` to the `iplogo.mr` file.

Of course, you may generate a bootstrap and write the converted image
on the disk at the same time:

	makeip -l iplogo.png -s iplogo.mr -v IP.BIN

## Information about some specific fields

Some fields used in the bootstrap need to be detailed, as they are have
constraints:

* Area Symbols
* Device Information
* Peripherals
* Release Date
* (Product) Version

### Area Symbols

The **Area Symbols** field consists of 8 characters, which are either space or
a specific letter. Each of these represent a geographical region in which the
disc is designed to work. So far, only the first 3 are assigned, the 5 others
were never been used so they are just left blank.

Supported area are **Japan** (and the rest of East Asia), **USA + Canada**, and
**Europe**.

If the character for a particular region is a space, the disc will not be
playable in that region. If it contains the correct region character, it will
be.

The region characters used are `J`, `U`, and `E`, respectively. So a disc only
playable in Europe would have an Area Symbols string of `"  E     "`. This
operation is made by **IP creator**.

To summarize, if you want your program region-free, you just have to assign
`JUE` to the **Area Symbols** field. This is the case by default.

### Device Information

The **Device Information** field begins with a 4 digit hexadecimal number,
which is a **CRC** on the **Product Number** and **Product Version** fields
(16 bytes). Then comes the string `" CD-ROM"`, and finally an indication of how
many discs this software uses, and which of these discs that this is. This is
indicated by 2 positive numbers separated with a slash. So if this is the second
disc of three, the **Device Information** field might be something like 
`"8B40 CD-ROM2/3  "`.

Please note, commercial `IP.BIN` files (which aren't produced by **IP creator**)
supports the `" GD-ROM"` type. Also, the **CRC** value doesn't need to be
provided as it's computed automatically by the **IP creator** program.

In clear, you just have to pass `CD-ROMx/y` to that field, where `x` is the disc
number and `y` the total discs in the set. The default value is `CD-ROM1/1`.

### Peripherals

The **Peripherals** field is a 28 bit long bitfield represented by a 7 digit
hexadecimal number. The meaning of the individual bits in each digit is given
below:

	<A><-------B------> <C->
	0000 0000 0000 0000 0000 0000 0000
	^^^^ ^^^^ ^^^^ ^^^^ ^^^^    ^    ^
	|||| |||| |||| |||| ||||    |    |
	|||| |||| |||| |||| ||||    |    +----- Uses Windows CE
	|||| |||| |||| |||| ||||    |
	|||| |||| |||| |||| ||||    +-----  VGA box support
	|||| |||| |||| |||| ||||
	|||| |||| |||| |||| |||+----- Other expansions
	|||| |||| |||| |||| ||+----- Puru Puru pack
	|||| |||| |||| |||| |+----- Mike device
	|||| |||| |||| |||| +----- Memory card
	|||| |||| |||| |||+------ Start + A + B + Directions
	|||| |||| |||| ||+------ C button
	|||| |||| |||| |+------ D button
	|||| |||| |||| +------ X button
	|||| |||| |||+------- Y button
	|||| |||| ||+------- Z button
	|||| |||| |+------- Expanded direction buttons
	|||| |||| +------- Analog R trigger
	|||| |||+-------- Analog L trigger
	|||| ||+-------- Analog horizontal controller
	|||| |+-------- Analog vertical controller
	|||| +-------- Expanded analog horizontal
	|||+--------- Expanded analog vertical
	||+--------- Gun
	|+--------- Keyboard
	+--------- Mouse

The group of bits indicated by `B` above indicate the discs minimum controller
requirements. So if the **Z button** bit is set, the software can not be used
with a controller that doesn't have a **Z button**. The `A` group indicates
which optional peripherals the disc supports. The `C` group indicates which
kinds of expansion units that the disc supports. The `VGA box support` bit
indicates that the disc can be run in **VGA** mode. The `WinCE` bit speaks for
itself and shouldn't be used for any homebrew program, so you have to leave this
field to `0`.

### Release Date

The **Release Date** field is in the `YYYYMMDD` format. By default, it's
initialized with today.

### (Product) Version

The **Version** field is the `Vx.yyy` format, where `x` is the major number and
`yyy` the minor number. Default is `V1.000`.

## GIMP MR Image Plug-In

[GIMP](https://www.gimp.org/) is a cross-platform image editor available for
GNU/Linux, macOS and Windows. The provided `gimp/file-mr.py` file is a 
[GIMP](https://www.gimp.org/) plug-in used for encode/decode **MR images**.
  
So by using this plug-in, you can load and save `.mr` files directly in GIMP.
As this plugin uses **Python-Fu**, you should have **Python extensions** enabled
in GIMP. [Read more here](https://docs.gimp.org/en/gimp-filters-python-fu.html).

To install the GIMP Plug-In:

1. Make the plug-in executable: ```chmod +x file-mr.py``` or equivalent.
2. Place the plug-in in the `plugins` directory. This directory is different
   depending on what operating system you use. To find out go to `GIMP` >
   `Preferences` > `Folders` (*Expand option*) > `Plugins` (see below an
   example under **Windows**, but this applies to others OS).
3. That's it!

![GIMP Plug-In Directory](img/gimp.png "GIMP Plug-In Directory")

## Acknowledgments

* [Jacob Alberty](https://github.com/jacobalberty) (LiENUS): Copyright-free
  bootstrap replacement (`IP.TMPL`). It is used by default in the **IP creator**
  program.
* [Marcus Comstedt](http://mc.pp.se/dc/sw.html)
  ([zeldin](https://github.com/zeldin)): Initial release of the 
  [IP creator](http://mc.pp.se/dc/sw.html) tool (`makeip`).
* [Andrew Kieschnick](http://napalm-x.thegypsy.com/andrewk/dc/) (ADK/Napalm):
  Code to insert `MR` images into `IP.BIN` and `PNG` to `MR` from
  [logotools](http://napalm-x.thegypsy.com/andrewk/dc/).
* [Hayden Kowalchuk](https://twitter.com/HaydenKowalchuk)
  ([mrneo240](https://github.com/mrneo240)): Fixes on the original `logotools`
  package.
* [Andress Antonio Barajas](https://twitter.com/bbhoodsta)
  ([BBHoodsta](https://gitlab.com/BBHoodsta)): Tons of fixes and improvements.
  Maker of the **GIMP MR Image Plug-in**.
* [SiZiOUS](http://sizious.com/): Complete refactoring of the project and
  provider of the snippet for decoding `MR` images (extracted from
  [Selfboot Inducer](https://sizious.com/download/dreamcast/#selfboot-inducer)).

## License

This project is licensed under the **BSD 2-clause "Simplified" License** - see
the [LICENSE](LICENSE) file for details.

CHANGELOG.md
# IP creator (makeip): Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [2.0.0] - 2020-06-24
### Added
- All the fields may now be filled directly from the command-line. Use the `-u`
  switch to get the list of all the available fields.
- All the fields are now validated when necessary (previously, only the **Area
  Symbols** field was validated).
- **PNG** images are now supported for logo insertion (extracted from `logotools`).
  **IP creator** may be used to convert a PNG image into a MR image by using
  the `-l` and `-s` switches together.

### Changed
- Complete redesign/refactoring of the whole code. Splitting code into many
  small source files instead of a single `makeip.c` file.
- Command-line options are now better handled.  
- The possibility to use an external `IP.TMPL` was restored by using the `-t`
  switch.
- The embedded `IP.TMPL` file is now the **LiENUS** version instead of the
  original file (which was using the copyrighted `AIP` library). 
- The `ip.txt` template file was changed a bit: the **Hardware ID** and 
  **Maker ID** fields aren't necessary anymore. The **Device Info** field
  doesn't need to contains the fake `0000` **CRC**, only the `CD-ROM1/1` value
  is needed now.  
  
## [1.5.0] - 2019-12-10
### Added
- Support of command-line arguments to edit the fields of the `IP.BIN`.
- Support of **MR image** insertion into the `IP.BIN`.

### Changed
- The `IP.TMPL` file is now embedded, it no longer depends of the external 
  `IP.TMPL` file.
- Program is now licensed under the **BSD 2-Clause License** (previously it
  wasn't licensed).

## [1.0.0] - 2000-09-05
### Added
- Initial (unversioned) release.


src/field.c
/* IP creator (makeip)
 *
 * Copyright (C) 2000, 2001, 2002, 2019, 2020 The KOS Team and contributors.
 * All rights reserved.
 *
 * This code was contributed to KallistiOS (KOS) by Andress Antonio Barajas
 * (BBHoodsta). It was originally made by Marcus Comstedt (zeldin). Some
 * portions of code were made by Andrew Kieschnick (ADK/Napalm). Heavily
 * updated by SiZiOUS. Bootstrap replacement (IP.TMPL) was made by Jacob
 * Alberty (LiENUS).
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice,
 *    this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#include "field.h"

field_t fields[NUM_FIELDS] = {
  { HARDWARE_ID,   "Hardware ID",   "SEGA SEGAKATANA",   0x0, 0x10, _check_fixed },
  { MAKER_ID,      "Maker ID",      "SEGA ENTERPRISES", 0x10, 0x10, _check_fixed },
  { DEVICE_INFO,   "Device Info",   "0000 CD-ROM1/1",   0x20, 0x10, _check_deviceinfo },
  { AREA_SYMBOLS,  "Area Symbols",  "JUE",              0x30,  0x8, _check_areasym },
  { PERIPHERALS,   "Peripherals",   "E000F10",          0x38,  0x8, _check_peripherals },
  { PRODUCT_NO,    "Product No",    "T-00000",          0x40,  0xa, NULL },
  { VERSION,       "Version",       "V1.000",           0x4a,  0x6, _check_version },
  { RELEASE_DATE,  "Release Date",  NULL,               0x50, 0x10, _check_date },
  { BOOT_FILENAME, "Boot Filename", "1ST_READ.BIN",     0x60, 0x10, NULL },
  { SW_MAKER_NAME, "SW Maker Name", "KallistiOS",       0x70, 0x10, NULL },
  { GAME_TITLE,    "Game Title",    "GAMETITLE",        0x80, 0x80, NULL },
};

int g_field_error = 0;

char *field_values[NUM_FIELDS];

void
init_release_date()
{
  time_t now;
  struct tm ts;
  char tmpbuf[0x10];

  time(&now);
  ts = *localtime(&now);
  strftime(tmpbuf, sizeof(tmpbuf), "%Y%m%d", &ts); // YYYYMMDD

  field_set_value(RELEASE_DATE, tmpbuf);
}

void
field_initialize()
{
  // initialize field values
  for (int i = 0; i < NUM_FIELDS; i++) {
    if (fields[i].default_value != NULL) {
      field_set_value(i, fields[i].default_value);
    }
  }

  // compute default release date
  init_release_date();
}

void
field_finalize()
{
  for (int i = 0; i < NUM_FIELDS; i++) {
    free(field_values[i]);
  }
}

char *
field_get_value(int index)
{
  return field_values[index];
}

char *
field_get_pretty_value(int index)
{
  char *result = field_get_value(index);
  char *deviceinfo = NULL;

  switch (index) {
    case AREA_SYMBOLS:
      trim(result);
      break;
    case DEVICE_INFO:
      deviceinfo = strchr(result, ' ');
      result = deviceinfo + 1; // skip the space char
      break;
  }

  return result;
}

int
field_set_value(int index, char *value)
{
  field_t *f = &fields[index];

  // checking the value length for that field
  if(strlen(value) > f->length) {
    log_error("data for field \"%s\" is too long\n", f->name);
    return 0;
  }

  // destroying previous allocated memory if needed
  if (field_values[index] != NULL) {
    free(field_values[index]);
  }

  // allocating the required memory
  field_values[index] = (char *) malloc((f->length + 1) * sizeof(char));
  memset(field_values[index], ' ', f->length);
  field_values[index][f->length] = '\0';

  // storing the value
  strcpy(field_values[index], value);

  // do additional checks if required
  if(f->extra_check != NULL && !(*f->extra_check)(f, field_values[index])) {
    g_field_error = 1;
    return 0;
  }

  log_notice("setting field \"%s\" to \"%s\"\n", f->name, field_get_pretty_value(index));

  return 1;
}

int
parse_file(FILE *fh)
{
  static char buf[80];
  int i, line = 0;

  while(fgets(buf, sizeof(buf), fh)) {
    char *p;

    line++;
    trim(buf);

    if(*buf) {
      if((p = strchr(buf, ':'))) {
        *p++ = '\0';
        trim(buf);

	for(i = 0; i < NUM_FIELDS; i++) {
          if(!strcmp(buf, fields[i].name))
            break;
        }

        if(i >= NUM_FIELDS) {
          log_error("unknown field \"%s\"\n", buf);
          return 0;
        }

        while(*p == ' ' || *p == '\t')
          p++;

        if (!field_set_value(i, p))
          return 0;

      } else {
        log_error("missing colon (\":\") on line %d\n", line);
        return 0;
      }
    }
  }

  return 1;
}

void
field_load(char *in)
{
  FILE *fh = fopen(in, "r");

  if(fh == NULL) {
    halt("can't open template: \"%s\"\n", in);
  }

  log_notice("loading template \"%s\"\n", in);

  int result;
  result = parse_file(fh);

  fclose(fh);

  if (!result) {
    exit(EXIT_FAILURE);
  }
}

void
field_write(char *ip)
{
  for (int i = 0; i < NUM_FIELDS; i++) {
    memset(ip + fields[i].position, ' ', fields[i].length);
    char *p = field_get_value(i);
    memcpy(ip + fields[i].position, p, strlen(p));
  }
}

int
field_erroneous() {
  return g_field_error;
}

int
_check_areasym(field_t *f, char *value)
{
  int i, a = 0;

  for(i = 0; i < strlen(value); i++) {
    switch(value[i]) {
     case 'J':
       a |= (1<<0);
       break;
     case 'U':
       a |= (1<<1);
       break;
     case 'E':
       a |= (1<<2);
       break;
     case '\0':
     case ' ':
       break;
     default:
       log_error("field \"%s\" contains an unknown area symbol '%c'\n", f->name, value[i]);
       return 0;
    }
  }

  for(i = 0; i < f->length; i++) {
    if((a & (1 << i)) == 0)
      value[i] = ' ';
    else
      value[i] = "JUE"[i];
  }

  return 1;
}

int
_check_date(field_t *f, char *value)
{
  int is_date = is_valid_date(value);

  if (!is_date) {
    log_error("field \"%s\" is invalid date (format is \"YYYYMMDD\")\n", f->name);	  
  }

  return is_date;
}

int
_check_fixed(field_t *f, char *value)
{
  int result = !(strcmp(f->default_value, value));

  if (!result) {
    log_error("field \"%s\" is not editable (must be \"%s\")\n", f->name, f->default_value);	  
  }

  return result;
}

int
_check_version(field_t *f, char *value)
{
  // value should be Vx.yyy
  int result;
  long major, minor;

  result = (strlen(value) == 6) && (value[0] == 'V') && (value[2] == '.') &&
    substr_long_parse(value, 1, 1, &major) &&
    substr_long_parse(value, 3, 3, &minor);

  if (!result) {
    log_error("field \"%s\" is invalid version (must be Vx.yyy)\n", f->name);
  }

  return result;
}

int
_check_peripherals(field_t *f, char *value)
{
  // value should be from '0000000' to 'FFFFF11'
  int result = 0;

  if (strlen(value) == 7) {
    // check peripherals
    char buf[6];
    strncpy(buf, value, 5);
    buf[5] = '\0';

    // check WinCE (useless but supported...) and VGA flags
    result = is_valid_hex(buf) && is_strict_bool(value[5]) &&
      is_strict_bool(value[6]);
  }

  if (!result) {
    log_error("field \"%s\" contains invalid values\n", f->name);
  }

  return result;
}

int
_check_deviceinfo(field_t *f, char *value)
{
  // can be legacy value like '0000 CD-ROMx/y' or 'CD-ROMx/y'
  int result = 1;

  char *deviceinfo = strdup(value);

  if (strlen(deviceinfo) == 14) { // '0000 CD-ROMx/y'
    char *device = strchr(deviceinfo, ' ');
    *device++ = '\0'; // deviceinfo = '0000'; device = 'CD-ROMx/y'
    result = result && (strlen(deviceinfo) == 4) && is_valid_hex(deviceinfo);
    memmove(deviceinfo, device, strlen(device)); // deviceinfo = 'CD-ROMx/y'
    deviceinfo[strlen(device)] = '\0';
  }

  long dummy;
  result = result && (strlen(deviceinfo) == 9)
    && ((!strncmp(deviceinfo, "CD-ROM", 6)) ||
        (!strncmp(deviceinfo, "GD-ROM", 6)))
    && substr_long_parse(deviceinfo, 6, 1, &dummy) && (deviceinfo[7] == '/')
	&& substr_long_parse(deviceinfo, 8, 1, &dummy);

  if (!result) {
    log_error("field \"%s\" contains invalid values (must be CD-ROMx/y)\n", f->name);
  } else if (strlen(deviceinfo) == 9) {
    // if short form (i.e. "CD-ROMx/y" only), append "0000" before
    char *buf = (char *) malloc(f->length * sizeof(char));
    strcpy(buf, "0000 "); // fake CRC that will be updated by makeip
    strcat(buf, deviceinfo); // append CD-ROMx/y
    strcpy(value, buf); // result
    free(buf);
  }

  free(deviceinfo);

  return result;
}
